lib/fluent/plugin/buffer.rb

```diff
@@ -793,19 +799,29 @@ def write_step_by_step(metadata, data, format, splits_count, &block)
                 else
                   chunk.append(split, compress: @compress)
                 end
+                adding_bytes = chunk.bytesize - committed_bytesize
 
                 if chunk_size_over?(chunk) # split size is larger than difference between size_full? and size_over?
-                  adding_bytes = chunk.instance_eval { @adding_bytes } || "N/A" # 3rd party might not have 'adding_bytes'
                   chunk.rollback
-
+                  committed_bytesize = chunk.bytesize
                   if split.size == 1 && original_bytesize == 0
-                    # It is obviously case that BufferChunkOverflowError should be raised here,
-                    # but if it raises here, already processed 'split' or
-                    # the proceeding 'split' will be lost completely.
-                    # so it is a last resort to delay raising such a exception
-                    errors << "concatenated/appended a #{adding_bytes} bytes record (nth: #{writing_splits_index}) is larger than buffer chunk limit size (#{@chunk_limit_size})"
-                    writing_splits_index += 1
-                    next
+                    if adding_bytes < @chunk_limit_size
+                      # As already processed content is kept after rollback, then unstaged chunk should be queued.
```

Comment:
I'm not sure but I think this block should be placed outer of `if split.size == 1 && original_bytesize == 0` block because probably this case can be occurred also on `split.size != 1` or `original_bytesize != 0`.

```diff
diff --git a/lib/fluent/plugin/buffer.rb b/lib/fluent/plugin/buffer.rb
index 7275d70c..f113f06b 100644
--- a/lib/fluent/plugin/buffer.rb
+++ b/lib/fluent/plugin/buffer.rb
@@ -804,24 +804,23 @@ module Fluent
                 if chunk_size_over?(chunk) # split size is larger than difference between size_full? and size_over?
                   chunk.rollback
                   committed_bytesize = chunk.bytesize
-                  if split.size == 1 && original_bytesize == 0
-                    if adding_bytes < @chunk_limit_size
-                      # As already processed content is kept after rollback, then unstaged chunk should be queued.
-                      # After that, re-process current split again.
-                      # New chunk should be allocated, to do it, modify @stage and so on.
-                      synchronize { @stage.delete(modified_metadata) }
-                      staged_chunk_used = false
-                      chunk.unstaged!
-                      break
-                    else
-                      # It is obviously case that BufferChunkOverflowError should be raised here,
-                      # but if it raises here, already processed 'split' or
-                      # the proceeding 'split' will be lost completely.
-                      # so it is a last resort to delay raising such a exception
-                      errors << "concatenated/appended a #{adding_bytes} bytes record (nth: #{writing_splits_index}) is larger than buffer chunk limit size (#{@chunk_limit_size})"
-                      writing_splits_index += 1
-                      next
-                    end
+                  if split.size == 1 && adding_bytes > @chunk_limit_size
+                    # It is obviously case that BufferChunkOverflowError should be raised here,
+                    # but if it raises here, already processed 'split' or
+                    # the proceeding 'split' will be lost completely.
+                    # so it is a last resort to delay raising such a exception
+                    errors << "concatenated/appended a #{adding_bytes} bytes record (nth: #{writing_splits_index}) is larger than buffer chunk limit size (#{@chunk_limit_size})"
+                    writing_splits_index += 1
+                    next
+                  end
+                  if adding_bytes < @chunk_limit_size
+                    # As already processed content is kept after rollback, then unstaged chunk should be queued.
+                    # After that, re-process current split again.
+                    # New chunk should be allocated, to do it, modify @stage and so on.
+                    synchronize { @stage.delete(modified_metadata) }
+                    staged_chunk_used = false
+                    chunk.unstaged!
+                    break
                   end
 
                   if chunk_size_full?(chunk) || split.size == 1
```

In addition, note that above patch includes the following change for detecting 1 big record:

```diff
- if split.size == 1 && original_bytesize == 0
+ if split.size == 1 && adding_bytes > @chunk_limit_size
```

I think it's more proper way to detect it.
(Without it, newly added tests are failed.)

Reply:
> because probably this case can be occurred also on `split.size != 1` or `original_bytesize != 0`.

Hmm, probably I was wrong. It seems that it doesn't occur when `split.size != 1`.

Reply:
I made it more understandable (for me...)